src/main/java/com/avispl/symphony/dal/avdevices/encoderdecoder/haivision/x4decoder/HaivisionX4DecoderCommunicator.java

```diff
@@ -0,0 +1,755 @@
+						SRT srt = streamStats.getSrt();
+
+						// Stream name filtering
+						if (this.streamName != null && setStreamName.remove(streamInfo.getName())) {
+							updateStreamStats(stats, streamInfo, streamStats, srt);
+							continue;
+						}
+
+						// Stream status filtering
+						if (this.streamStatus != null && !setStreamStatus.remove(streamStats.getState())) {
+							continue;
+						}
+
+						// Port number filtering
+						if (this.portNumber != null) {
+							Integer port = Integer.parseInt(streamInfo.getPort());
+							boolean isValidPort = handleAdapterPortRangeFromUser(port, setPortNumber);
+							if (!isValidPort) {
+								continue;
+							}
+						}
+						updateStreamStats(stats, streamInfo, streamStats, srt);
+					}
+
+					// Update invalid streams name
+					if (setStreamName != null) {
+						updateInvalidAdapterProperties(stats, setStreamName, FilteringErrorGroup.STREAM_NAME.getName(), StreamMonitoringMetric.NAME.getName());
+					}
+					// Update invalid stream status
+					if (setStreamStatus != null) {
+						updateInvalidAdapterProperties(stats, setStreamStatus, FilteringErrorGroup.STREAM_STATUS_AND_PORT_NUMBER.getName(), StreamMonitoringMetric.STATE.getName());
+					}
+					// Update invalid port number
+					if (setPortNumber != null) {
+						updateInvalidAdapterProperties(stats, setPortNumber, FilteringErrorGroup.STREAM_STATUS_AND_PORT_NUMBER.getName(), StreamConfigMetric.PORT.getName());
+					}
+				} else {
+					updateStreamStatisticsFailedMonitor(failedMonitor);
+				}
+			} else {
+				updateStreamStatisticsFailedMonitor(failedMonitor);
+			}
+		} catch (Exception e) {
+			updateStreamStatisticsFailedMonitor(failedMonitor);
+		}
+	}
+
+	/**
+	 * Counting metric group is failed to monitor
+	 *
+	 * @return number failed monitoring metric group in the metric
+	 */
+	private int getNoOfFailedMonitorMetricGroup() {
+		int noOfFailedMonitorMetric = 3;
+		for (MonitoringMetricGroup metric : MonitoringMetricGroup.values()) {
+			if (metric.isFailedMonitorCheck()) {
+				noOfFailedMonitorMetric++;
+			}
+		}
+		return noOfFailedMonitorMetric;
+	}
+
+
+	/**
+	 * This method is used to populate all monitoring properties:
+	 * <li>Decoders statistic</li>
+	 * <li>Streams statistic</li>
+	 *
+	 * @param stats list statistic property
+	 * @throws ResourceNotReachableException when failedMonitor said all device monitoring data are failed to get
+	 */
+	private void populateDecoderMonitoringMetrics(Map<String, String> stats) {
+		Objects.requireNonNull(stats);
+
+		if (!StringUtils.isNullOrEmpty(getPassword()) && !StringUtils.isNullOrEmpty(getLogin())) {
+			retrieveSessionFromDecoder();
+			retrieveUserRoleBasedFromDecoder();
+		} else {
+			this.authenticationCookie.setSessionID(DecoderConstant.AUTHORIZED);
+		}
+		// Retrieving all device info
+		retrieveDeviceInfo(stats);
+
+		// Retrieving all decoders stats
+		for (int decoderID = 0; decoderID < 4; decoderID++) {
```

Comment:
instead of 0 and 4, I would call it as constant, MIN_DECODER_INDEX, MAX_DECODER_INDEX, etc.

Reply:
I fixed it